test-e2e/test_zookeeper_backup.py

```diff
@@ -0,0 +1,283 @@
+"""
+Tests for the ZooKeeper backup/restore via `dcos-zk`.
+"""
+
+import logging
+import uuid
+from pathlib import Path
+from typing import Set
+
+import pytest
+from _pytest.fixtures import SubRequest
+
+from cluster_helpers import wait_for_dcos_oss
+from dcos_e2e.backends import Docker
+from dcos_e2e.cluster import Cluster
+from dcos_e2e.node import Node, Output
+from kazoo.client import KazooClient
+from kazoo.client import KazooState
+from kazoo.exceptions import NoNodeError
+
+
+LOGGER = logging.getLogger(__name__)
+
+# Arbitrary value written to ZooKeeper.
+FLAG = b'flag'
+
+
+@pytest.fixture
+def three_master_cluster(
+    artifact_path: Path,
+    docker_backend: Docker,
+    request: SubRequest,
+    log_dir: Path,
+) -> Cluster:
+    """
+    Spin up a highly-available DC/OS cluster with three master nodes.
+    """
+    with Cluster(
+        cluster_backend=docker_backend,
+        masters=3,
+        agents=0,
+        public_agents=0,
+    ) as cluster:
+        cluster.install_dcos_from_path(
+            dcos_installer=artifact_path,
+            dcos_config=cluster.base_config,
+            ip_detect_path=docker_backend.ip_detect_path,
+        )
+        wait_for_dcos_oss(
+            cluster=cluster,
+            request=request,
+            log_dir=log_dir,
+        )
+        yield cluster
+
+
+@pytest.fixture
+def zk_client(three_master_cluster: Cluster) -> KazooClient:
+    """
+    ZooKeeper client connected to a given DC/OS cluster.
+    """
+    zk_hostports = ','.join(['{}:2181'.format(m.public_ip_address)
+                             for m in three_master_cluster.masters])
+    zk_client = KazooClient(
+        hosts=zk_hostports,
+        # Avoid failure due to client session timeout.
+        timeout=60,
+    )
+    def zk_listener(state: KazooState) -> None:
+        # Open session after temporary connection loss.
+        if state == KazooState.CONNECTED:
+            zk_client.start()
+
+    zk_client.add_listener(zk_listener)
```

Comment:
Why do we need this? The docs at https://kazoo.readthedocs.io/en/latest/basic_usage.html#connection-handling state that:
```
Once connected, the client will attempt to stay connected regardless of intermittent connection loss or Zookeeper session expiration.
```

Reply:
`Connected` doesn't seem to imply able to do `get()` operations. I'm still in the progress of figuring this out.

Reply:
Actually, with a long enough connection timeout we don't seem to need this.